### PR TITLE
Add GitBucket support (refs #142)

### PR DESCRIPTION
**Pull request by kounoike**
[GitBucket](https://github.com/gitbucket/gitbucket) is Git platform with high-API compatibility to GitHub (for client, GitBucket looks like as GitHub Enterprise).
This PR add GitBucket support with some actions.

There are few changes to git-repo side.

- add support to ssh:// style remote.
- also support http, not only https.
- Change refspec in github.py:request_fetch

gitbucket.py now supports

- login with user created access token.
- list repositories
- clone repository
- fetch PR

these action doesn't supported now.

- automatic get access token by `git repo config`
- delete repository
- create PR
- list PR

### CODE REVIEW

**Comment by guyzmo** [comment]
hi @kounoike I'm starting to review your code, and I'd appreciate if you could split it in a four commits, one per file. Because it's likely I won't merge your PR as is, but rather gradually (the fixes first, and the gitbucket after), so I'd like to be able to preserve attribution to you for your changes 😉

**Comment by kounoike** [comment]
@guyzmo Thanks to review! OK, I will split my commit. But, I'm not accustomed to such non-standard Git operations. I split to four commit locally. How to push it for GitHub? `git push origin` or `git push -f origin`?

**Comment by guyzmo** [comment]
Thank you!

You should `git push -f github pr-gitbucket` (or `origin` instead of `github`).
It's OK to rewrite history into a branch that servers as a PR.

**Comment by kounoike** [comment]
@guyzmo OK. I pushed it!

File: git_repo/services/service.py [hunk 1]
```diff
@@ -244,7 +244,10 @@ def format_path(self, repository, namespace=None, rw=False):
         if not rw and '/' in repo:
             return '{}/{}'.format(self.url_ro, repo)
         elif rw and '/' in repo:
-            return '{}:{}'.format(self.url_rw, repo)
+            if self.url_rw.startswith('ssh://'):
```
Comment (guyzmo) [inline git_repo/services/service.py]: in what context do you need `ssh://` URIs? As far as I can tell, it's not supported by neither Git or SSH directly (only by `xdg-` tools or browsers.

Comment (kounoike) [inline git_repo/services/service.py]: GitBucket's ssh url is:  `ssh://git@server:29418/root/repo.git`

GitBucket doesn't use system sshd. It uses embedded sshd. So, it can't listen on 22 port.
By [Git manual](https://git-scm.com/docs/git-clone#_git_urls_a_id_urls_a), "An alternative scp-like syntax" doesn't take ssh port. So it is required for GitBucket with ssh repository access.

Comment (guyzmo) [inline git_repo/services/service.py]: Interesting, that I did not know.

Comment (guyzmo) [inline git_repo/services/service.py]: could you update the commits related to that change the title of the commits to 'Adds missing ssh:// URL support' and in the body the details above? It's not only for GitBucket support, it's actually an important fix 😉

Comment (kounoike) [inline git_repo/services/service.py]: I fixed it now. `push -f` again?

Comment (guyzmo) [inline git_repo/services/service.py]: yup 👍

Comment (kounoike) [inline git_repo/services/service.py]: pushed!

File: git_repo/repo.py [hunk 2]
```diff
@@ -163,15 +163,16 @@ def _guess_repo_slug(self, repository, service, resolve_targets=None):
...
-                    if url.startswith('https'):
-                        if url.endswith('.git'):
-                            url = url[:-4]
+                    if url.endswith('.git'):
+                        url = url[:-4]
+                    if url.startswith('https') or url.startswith('http'):
                         *_, user, name = url.split('/')
                         self.set_repo_slug('/'.join([user, name]))
                         return
+                    elif url.startswith('ssh://'):
```
Comment (guyzmo) [inline git_repo/repo.py]: instead of repeating the above code body below, you can simply write:

```python
if url.startswith('https') or url.startswith('http') or url.startswith('ssh'):
```

File: git_repo/repo.py [hunk 3]
```diff
@@ -163,15 +163,16 @@ def _guess_repo_slug(self, repository, service, resolve_targets=None):
...
+                        *_, user, name = url.split('/')
+                        self.set_repo_slug('/'.join([user, name]))
```
Comment (guyzmo) [inline git_repo/repo.py]: if we were keeping that body, the `return` statement is missing here, to break from the `for` loop. (technically, it's more a `break` than a `return`, but in this case they're the same).

File: git_repo/repo.py (hunk 2 above)
Comment (kounoike) [inline git_repo/repo.py]: OK I will fix it, but I think it is more better:
```python
if url.startswith('https://') or url.startswith('http://') or url.startswith('ssh://'):
```

because, "An alternative scp-like syntax" can use any username such as:

- `http@server:owner/repo.git`
- `https@server:owner/repo.git`
- `ssh@server:owner/repo.git`

I know these format can be used by Gogs. Gogs can configure ssh username. (Of course it is rare case)

Comment (kounoike) [inline git_repo/repo.py]: > Gogs can configure ssh username. (Of course it is rare case)

hmm... it means `startswith('git@')` doesn't match for Gogs...
In my Gogs instance running on Raspberry Pi, ssh URL is: `pi@raspberrypi.local:test/test.git`

But it is not GitBucket related fix. Which is better action?
Should I create new Issue/PR? or fix it in this PR?

It should be rewrite as:
```
                for url in remote.urls:
                    if url.endswith('.git'):
                        url = url[:-4]
                    if url.find('://') > -1:
                        *_, user, name = url.split('/')
                        self.set_repo_slug('/'.join([user, name]))
                        return
                    elif url.find('@') > -1:
                        _, repo_slug = url.split(':')
                        self.set_repo_slug(repo_slug)
                        return
```

but it is not tested.

File: git_repo/services/ext/gitbucket.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        super(GitbucketService, self).__init__(*args, **kwarg)
+
+    def connect(self):
+        self.gh = github3.GitHubEnterprise(self.build_url(self))
+        super(GitbucketService, self).connect()
+
+    @classmethod
+    def get_auth_token(cls, login, password, prompt=None):
+        print("Please edit _YOUR_ACCESS_TOKEN_ to actual token in ~/.gitconfig or ~/.git/config after.")
+        print("And add ssh-url = ssh://git@{}:_YOUR_SSH_PORT_".format(cls.fqdn))
```
Comment (guyzmo) [inline git_repo/services/ext/gitbucket.py]: the prompt parameter is actually the `loop_input` callback given from git-repo's CLI interface.

File: git_repo/repo.py (hunk 2 above)
Comment (guyzmo) [inline git_repo/repo.py]: well, I'm ok if you take all the SSH format fixes as another PR. 

I was anyway planning to merge them first before merging your gitbucket PR.

File: git_repo/services/ext/gitbucket.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+from ..service import register_target
+from .github import GithubService
+
+import github3
+
+@register_target('bucket', 'gitbucket')
+class GitbucketService(GithubService):
+    fqdn = "localhost"
+    port = 8080
```
Comment (kounoike) [inline git_repo/services/ext/gitbucket.py]: There is demo site at https://gitbucket.herokuapp.com/
But I don't know who maintains it, who pays money for heroku's charge. So I can't decide it to default.

File: git_repo/repo.py (hunk 2 above)
Comment (guyzmo) [inline git_repo/repo.py]: if you do another PR, make a branch for each PR and don't forget to rebase the branch for this PR on top of the other one! ☺

File: git_repo/services/ext/gitbucket.py (hunk 5 above)
Comment (guyzmo) [inline git_repo/services/ext/gitbucket.py]: And it's not opened for accounts. It'd be nice to find an almost open instance, like gogs do, to default to it for tests and all.

Comment (kounoike) [inline git_repo/services/ext/gitbucket.py]: try.gogs.io is down now...

Comment (guyzmo) [inline git_repo/services/ext/gitbucket.py]: it's nice that we recorded the tests before it went down 😁

File: git_repo/services/ext/gitbucket.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    def __init__(self, *args, **kwarg):
+        super(GitbucketService, self).__init__(*args, **kwarg)
+
+    def connect(self):
+        self.gh = github3.GitHubEnterprise(self.build_url(self))
```
Comment (kounoike) [inline git_repo/services/ext/gitbucket.py]: I want to replace `self.gh` object from `github3.GitHub` to `github3.GithHubEnterprise` with GitBucket's API endpoint URL.

File: git_repo/services/ext/gitbucket.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        #except github3.models.GitHubError as err:
+        #    if len(err.args) > 0 and 422 == err.args[0].status_code:
+        #        raise ResourceExistsError("A token already exist for this machine on your github account.")
+        #    else:
+        #        raise err
+
+    def format_path(self, repository, namespace=None, rw=False):
+        path = super(GitbucketService, self).format_path(repository, namespace, rw)
+        if not path.endswith(".git"):
+            path = "{}.git".format(path)
```
Comment (kounoike) [inline git_repo/services/ext/gitbucket.py]: I know it.

In GitBucket, 
for ssh access, `.git` is required.
for http(s) access, GitBucket's formal http(s) URL format is `http://server:port/git/owner/repo.git`. 
But it also can access with:

- `http://server:port/git/owner/repo`
- `http://server:port/owner/repo.git`

But it can't access by `http://server:port/owner/repo`. Then it always add `.git`.

Recently, I found problem with git-2.12. I wrote it in https://github.com/gitbucket/gitbucket/issues/1552.

In GitBucket-4.12, `http://server:port/owner/repo.git` returns redirect to `http://server:port/git/owner/repo.git`. git-2.12 doesn't allow redirect url.
 so I will rewrite `format_path` such as:
- never call `super(...).format_path`
- returns `ssh://user@server:port/owner/repo.git` for rw=True
- returns `http://server:port/git/owner/repo.git` for rw=False

Comment (guyzmo) [inline git_repo/services/ext/gitbucket.py]: ok, makes sense

File: git_repo/services/ext/gitbucket.py (hunk 6 above)
Comment (guyzmo) [inline git_repo/services/ext/gitbucket.py]: ok, but that's going to be an issue. The `self.gh` instance shall be created in the constructor as first line [even before the call the the constructor](https://github.com/guyzmo/git-repo/blob/devel/git_repo/services/ext/github.py#L20,L22).

The reason for that is the tests, where it pulls [the `session` instance out of the `self.gh` instance](https://github.com/guyzmo/git-repo/blob/devel/tests/integration/test_github.py#L35,L36). If you change the `self.gh` instance at `self.connect()` time, there's a high probability that the tests will be testing another session instance than the one you're actually using in your methods.

So, we could patch [GithubService.__init__()](https://github.com/guyzmo/git-repo/blob/devel/git_repo/services/ext/github.py#L21,L21) so that:

````python
GITHUB_FQDN='github.com'

@register_target('hub', 'github')
class GithubService(RepositoryService):
    fqdn = GITHUB_FQDN

    def __init__(self, *args, **kwarg):
        if self.fqdn == GITHUB_FQDN:
            self.gh = github3.GitHub()
       else:
            self.gh = github3.GitHubEnterprise(RepositoryService.build_url(self))
        super(GithubService, self).__init__(*args, **kwarg)
```

and as a side effect, there's a chance that actually it enables `GithubService` to work with self hosted github instances… out of the box 🙌

File: git_repo/repo.py (hunk 2 above)
Comment (kounoike) [inline git_repo/repo.py]: OK. I will create new PR titled as 'Add support for ssh:// URL and any ssh user on scp-style URL'.

hmm... rebase this PR's branch onto new one?

Comment (guyzmo) [inline git_repo/repo.py]: > *hmm... rebase this PR's branch onto new one?*

because your commits are already in order, actually you don't need to go into that much trouble. 

Just do:

```
% git checkout 460d27e19bf322c51eb3d7e03c53458830074b9c
% git co -b pr-sshpatch
% git push github pr-sshpatch # or your origin branch
% git hub request guyzmo/git-repo create # or click in the webui
```

Comment (kounoike) [inline git_repo/repo.py]: Thanks, I do it.
But it contains 'Change local branch name to remote repo name.' commit.
I think it is not related change for ssh:// URL things.

Comment (guyzmo) [inline git_repo/repo.py]: it's ok, it's a minor bug that also needs to get fixed 😉

Comment (kounoike) [inline git_repo/repo.py]: I create new PR #148 for `ssh://` and `user@server:pathto/repo.git` URL fixes. and I will remove these changes.

Comment (guyzmo) [inline git_repo/repo.py]: nah, you can keep this PR on top of the other PR it's ok, I'm merging it ASAP

Comment (guyzmo) [inline git_repo/repo.py]: just add the third commit you forgot 😉

**Comment by guyzmo** [comment]
can you rebase this branch on top of the devel branch of `guyzmo/git-repo`?

**Comment by guyzmo** [comment]
ok, so I've rebased this branch on devel in the branch `requests/github/144` on my repository. I'll need you to change the current branch with the one I've added, so I'll need you to do the following.

Let's consider the following two remotes:

* `github` points to your repository `kounoike/git-repo`
* `upstream` points to my repository `guyzmo/git-repo`

if it's not like that, adapt accordingly to what you chose

```
# first make sure you're in this PR's branch
git checkout pr-gitbucket
# then make a backup of it
git checkout -b pr-gitbucket-backup
# remove the branch
git branch -D pr-gitbucket
# fetch the branch I've pushed
git fetch upstream requests/github/144
# check out that branch
git checkout upstream/requests/github/144
# store that branch locally as the new pr-gitbucket
git checkout -b pr-gitbucket
# and force push it to your github project
git push -f github pr-gitbucket
# and now this PR is rebased on the new main devel branch
```

**Comment by guyzmo** [comment]
Once the above is done, I'll need you to:

* fix `__init__()`/`connect()` scheme
*  update `get_auth_token()` as I suggest
* add tests to this branch
  * and a deployment script for gitbucket would be nice to have, to replicate the tests when needed

and BTW, thank you for your care so far 🙌
and don't hesitate to join our IRC channel on freenode [#git-repo](https://webchat.freenode.net/?channels=#git-repo)

File: git_repo/services/ext/gitbucket.py (hunk 6 above)
Comment (kounoike) [inline git_repo/services/ext/gitbucket.py]: I try it. but it doesn't load fqdn/port from configuration on this phase. Because, load_configuration is called by `RepositoryService.__init__`. And `RepositoryService.__init__` also calls `self.connect`, so self.gh must be initialized before __init__ called.
@guyzmo do you have any idea?

Comment (guyzmo) [inline git_repo/services/ext/gitbucket.py]: ok, I'll make a ticket to github3.py for this issue, so he splits the init in several steps. (by having a default URL to an empty invalid value, and a `set_url()` method)


Currently, here's the [code it is doing](https://github.com/sigmavirus24/github3.py/blob/0.9.6/github3/github.py#L1490,L1493) at startup time.

So we keep my suggestion above, but the `connect()` method shall change:

```python
def connect(self):
    self.gh._session.base_url = url.rstrip('/') + '/api/v3'
    self.gh._session.verify = self.session_certificate or not self.session_insecure
    super(GitbucketService, self).connect()
```

Comment (kounoike) [inline git_repo/services/ext/gitbucket.py]: ...?
I talk about timing of loading settings from `.gitconfig`.

1. on `GitBucketService.__init__` calls `GithubService.__init__`
2. on `GithubService.__init__`, assign `self.gh` to `GitHubEnterprise` with `build_url`
3. on `GithubService.__init__`, calls `RepositoryService.__init__`
4. on `RepositoryService.__init__` calls `load_configuration`

`.gitconfig`'s fqdn/port settings load at 4. but `build_url` called at 2.

and even if rewrite `__init__` in `github.py` as:
```python
    def __init__(self, *args, **kwarg):
        super(GithubService, self).__init__(*args, **kwarg)```
        if self.fqdn == GITHUB_COM_FQDN:
            self.gh = github3.GitHub()
        else:
            self.gh = github3.GitHubEnterprise(RepositoryService.build_url(self))
```

it get error because `RepositoryService.__init__` calls `self.connect`. it uses `self.gh`.

Comment (guyzmo) [inline git_repo/services/ext/gitbucket.py]: cf [how I did it in gitlab](https://github.com/kounoike/git-repo/blob/e48a3b761102eb278351032a7c4b3fbc4c1034c2/git_repo/services/ext/gitlab.py#L24,L36).

The order is:

1. init phase:
  1. instanciate session (and stuff that go with it)
  2. load configuration
2. connect phase:
  1. if need be, overload the github reference (but preserve the session)
  2. do the connect stuff

basically here's the fixed gist:

```python
GITHUB_FQDN='github.com'

@register_target('hub', 'github')
class GithubService(RepositoryService):
    fqdn = GITHUB_FQDN

    def __init__(self, *args, **kwarg):
        self.gh = github3.GitHub()
        super(GithubService, self).__init__(*args, **kwarg)

    def connect(self):
        if self.fqdn != GITHUB_FQDN:
            # upgrade self.gh from a GitHub object to a GitHubEnterprise object
            gh = github3.GitHubEnterprise(RepositoryService.build_url(self))
            self.gh._session.base_url = gh._session.base_url
            self.gh = gh
            # propagate ssl certificate parameter
            self.gh._session.verify = self.session_certificate or not self.session_insecure
        try:
            self.gh.login(token=self._privatekey)
            self.username = self.gh.user().login
        except github3.models.GitHubError as err:
            if 401 == err.code:
                if not self._privatekey:
                    raise ConnectionError('Could not connect to Github. '
                                          'Please configure .gitconfig '
                                          'with your github private key.') from err
                else:
                    raise ConnectionError('Could not connect to Github. '
'Check your configuration and try again.') from err
```

File: git_repo/services/ext/github.py [hunk 8]
```diff
@@ -13,15 +13,24 @@
...
 
     def __init__(self, *args, **kwarg):
         self.gh = github3.GitHub()
         super(GithubService, self).__init__(*args, **kwarg)
 
     def connect(self):
+        if self.fqdn != GITHUB_COM_FQDN:
+            # upgrade self.gh from a GitHub object to a GitHubEnterprise object
+            gh = github3.GitHubEnterprise(RepositoryService.build_url(self))
+            self.gh._session.base_url = gh._session.base_url
```
Comment (guyzmo) [inline git_repo/services/ext/github.py]: this is wrong, it should be the other way around:

```python
            gh._session.base_url = self.gh._session.base_url
```

Comment (kounoike) [inline git_repo/services/ext/github.py]: OK I fixed it.

- self.gh must change to new one(GitHubEnterprise)
- self.gh._session must keeps old one(created by __init__)
- self.gh._session.base_url must change to new URL

```python
            gh = github3.GitHubEnterprise(RepositoryService.build_url(self))
            self.gh._session.base_url = gh._session.base_url
            gh._session = self.gh._session
            self.gh = gh
```
